Unify compute volume fractions

gassmoeller opened the pull request:
Closes #2121 and is a first step towards #1767. All of the unify_volume_fractions functions were the same (one minor deviation in visco_plastic and compositional_heating, that is why I introduced the component mask). Hopefully this stops the spreading of this function into ever more material models.

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -70,60 +70,31 @@ namespace aspect
-      // assign compositional fields associated with strain a value of 0
+      // Store which components to exclude during volume fraction computation.
+      ComponentMask composition_mask(this->n_compositional_fields(),true);
       if (use_strain_weakening == true)
```
What if the strain tensor is tracked, but not used for weakening? Is that not supported? (I know that technically this doesn't relate to your pull request...)

gassmoeller replied:
The `use_finite_strain_tensor` variable in this material model explicitly means use it for strain weakening. It is not meant for anything else.

jdannberg commented on include/aspect/material_model/visco_plastic.h:
```diff
@@ -160,8 +160,6 @@ namespace aspect
         double grain_size;
 
-        std::vector<double> compute_volume_fractions(
-          const std::vector<double> &compositional_fields) const;
```
Why are these all protected and not private?

jdannberg commented on source/material_model/visco_plastic.cc:
```diff
@@ -70,60 +70,31 @@ namespace aspect
-                    const enum averaging_scheme &average_type) const
-    {
-      double averaged_parameter = 0.0;
       const std::vector<double> volume_fractions = compute_volume_fractions(composition);
```
Shouldn't that get the `composition_mask`? I am surprised that it's working like this...

gassmoeller commented on benchmarks/nonlinear_channel_flow/simple_nonlinear.cc:
```diff
@@ -55,9 +55,6 @@ namespace aspect
       private:
-        std::vector<double> compute_volume_fractions( const std::vector<double> &compositional_fields) const;
-
-
```
there are a lot of nonlinear channel flow tests that use this material model

jdannberg replied:
Ah, okay :+1:

gassmoeller commented:
I think I addressed all comments.

bangerth commented on include/aspect/heating_model/compositional_heating.h:
```diff
@@ -99,7 +80,7 @@ namespace aspect
          * etc.
          */
-        std::vector<int> field_used_in_heat_production_averaging;
+        std::vector<bool> fields_used_in_heat_production_averaging;
```
Is it correct that the length of this array is now equal to the number of compositional fields? What was it before -- the number of fields that are used, with each vector entry an index between zero and the number of compositional fields?

bangerth commented on include/aspect/material_model/dynamic_friction.h:
```diff
@@ -1,5 +1,5 @@
 /*
-  Copyright (C) 2014 - 2018 by the authors of the ASPECT code.
+  Copyright (C) 2014 - 2017 by the authors of the ASPECT code.
```
You probably want to do undo this.

bangerth commented on include/aspect/material_model/dynamic_friction.h:
```diff
@@ -177,8 +163,8 @@ namespace aspect
-         * The dynamic coefficient of friction
-         */
+        * The dynamic coefficient of friction
+        */
```
and this

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -779,6 +780,25 @@ namespace aspect
+     * one, we assume that there is no background mantle (i.e., that field value
+     * is zero). Otherwise, the difference between the sum of the compositional
+     * fields and 1.0 is assumed to be the amount of background mantle.
+     * Optionally, one can input a component mask that determines, which of the
```
no comma after 'determines'

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -779,6 +780,25 @@ namespace aspect
+     * fields and 1.0 is assumed to be the amount of background mantle.
+     * Optionally, one can input a component mask that determines, which of the
+     * compositional fields to use during the computation (e.g. because
+     * some fields do contain non-volumetric quantities like strain,
```
I would drop 'do' here because it puts emphasis on something that I find misleading.

bangerth commented on source/heating_model/compositional_heating.cc:
```diff
@@ -136,16 +89,23 @@ namespace aspect
-                                                                   "Compositional heating values");
-
-
+          std::vector<int> used_fields = Utilities::possibly_extend_from_1_to_N (
```
I think you can make this `const`

bangerth commented on source/heating_model/compositional_heating.cc:
```diff
@@ -136,16 +89,23 @@ namespace aspect
+
+          fields_used_in_heat_production_averaging.resize(used_fields.size());
+          for (unsigned int i=0; i<used_fields.size(); ++i)
+            fields_used_in_heat_production_averaging[i] = used_fields[i];
```
Is this correct? The right hand side are integers, the left are booleans. If this is correct, then I misunderstood what `used_fields` contained before. Can you make the cast explicit by putting a boolean expression on the rhs?

gassmoeller commented:
I addressed the comments and while doing that found an existing bug in compositional heating that interfered with this PR. It would be great if somebody could take a look at the new changelog entry, the compositional heating plugin, and the new test compositional_heating_exclude_background to make sure I got this right. In essence there was an off-by-one bug in the use of the `fields_used_in_heat_production_averaging` vector. The first entry was supposed to be used for the background field, but was actually used for the first compositional field when computing volume fractions.

naliboff commented:
Ugh, thanks for catching that Rene. The test I originally ran just worked by proxy of having the right combination of values.
